features/lorawan/LoRaWANStack.h

```diff
@@ -0,0 +1,290 @@
+     *                          are reserved, whereas port numbers from 1 to 223
+     *                          (0x01 to 0xDF) are valid port numbers.
+     *                          Anything out of this range is illegal.
+     *
+     * @param data              A pointer to buffer where the received data will be
+     *                          stored.
+     *
+     * @param length            The size of data in bytes
+     *
+     * @param flags             A flag is used to determine what type of
+     *                          message is being received, for example:
+     *
+     *                          MSG_UNCONFIRMED_FLAG = 0x01,
+     *                          MSG_CONFIRMED_FLAG = 0x02
+     *                          MSG_MULTICAST_FLAG = 0x04,
+     *                          MSG_PROPRIETARY_FLAG = 0x08
+     *
+     *                          MSG_MULTICAST_FLAG and MSG_PROPRIETARY_FLAG can be
+     *                          used in conjunction with MSG_UNCONFIRMED_FLAG and
+     *                          MSG_CONFIRMED_FLAG depending on the intended use.
+     *
+     *                          MSG_PROPRIETARY_FLAG|MSG_CONFIRMED_FLAG mask will set
+     *                          a confirmed message flag for a proprietary message.
+     *
+     *                          MSG_CONFIRMED_FLAG and MSG_UNCONFIRMED_FLAG are
+     *                          not mutually exclusive, i.e., the user can subscribe to
+     *                          receive both CONFIRMED AND UNCONFIRMED messages at
+     *                          the same time.
+     *
+     * @return                  It could be one of these:
+     *                             i)   0 if there is nothing else to read.
+     *                             ii)  Number of bytes still pending to read.
+     *                             iii) LORA_MAC_STATUS_WOULD_BLOCK if there is
+     *                                  nothing available to read at the moment.
+     *                             iv)  A negative error code on failure.
+     */
+    int16_t handle_rx(const uint8_t port, uint8_t* data,
+                      uint16_t length, uint8_t flags);
+
+    /** Shuts down the LoRaWAN protocol.
+     *
+     * In response to the user call for disconnection, the stack shuts down itself.
+     */
+    void shutdown_lorawan();
+
+private:
+    LoRaWANStack();
+    ~LoRaWANStack();
+
+    bool is_port_valid(uint8_t port);
+    lora_mac_status_t lora_state_machine();
+    void set_device_state(device_states_t new_state);
+    int16_t prepare_special_tx_frame(uint8_t port);
+    lora_mac_status_t send_frame_to_mac();
+    lora_mac_region_t get_lora_phy();
+    void on_tx_next_packet_timer_event(void);
+    void mcps_confirm(McpsConfirm_t *mcps_confirm);
+    void mcps_indication(McpsIndication_t *mcps_indication);
+    void mlme_confirm(MlmeConfirm_t *mlme_confirm);
+
+    lora_mac_status_t mlme_request_handler(lora_mac_mlme_req_t *mlme_request);
```

Comment:
Should private methods be  documented also? Might help developers.

Reply:
Yeah, I will put something there.